File: frontend/BusinessFlow/src/app/modules/crm/services/opportunity.service.ts
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { ApiService, PagedResponse } from '../../../core/services/api.service';
import { ChangeStageRequest, DuplicateMatch, Opportunity, OpportunityStage, PipelineSummary } from '../models/crm.model';

@Injectable({ providedIn: 'root' })
export class OpportunityService {
  private readonly endpoint = '/crm/opportunities';

  constructor(private api: ApiService) {}

  list(page = 0, size = 20, params?: any): Observable<PagedResponse<Opportunity>> {
    return this.api.getPaged<Opportunity>(this.endpoint, page, size, params);
  }

  getById(id: number): Observable<Opportunity> {
    return this.api.get<Opportunity>(`${this.endpoint}/${id}`);
  }

  create(payload: Partial<Opportunity>): Observable<Opportunity> {
    return this.api.post<Opportunity>(this.endpoint, payload);
  }

  createFromLead(leadId: number, payload: Partial<Opportunity>): Observable<Opportunity> {
    return this.api.post<Opportunity>(`${this.endpoint}/from-lead/${leadId}`, payload);
  }

  update(id: number, payload: Partial<Opportunity>): Observable<Opportunity> {
    return this.api.patch<Opportunity>(`${this.endpoint}/${id}`, payload);
  }

  changeStage(id: number, stage: OpportunityStage, options?: Partial<ChangeStageRequest>): Observable<Opportunity> {
    return this.api.patch<Opportunity>(`${this.endpoint}/${id}/stage`, { stage, ...options });
  }

  // Called before committing a WON stage change on a client-less opportunity, so the UI
  // can confirm link-existing vs create-new before the transition happens.
  previewWonDuplicate(id: number): Observable<DuplicateMatch | null> {
    return this.api.get<DuplicateMatch | null>(`${this.endpoint}/${id}/won-duplicate-check`);
  }

  pipelineSummary(): Observable<PipelineSummary> {
    return this.api.get<PipelineSummary>(`${this.endpoint}/pipeline-summary`);
  }

  delete(id: number): Observable<void> {
    return this.api.delete<void>(`${this.endpoint}/${id}`);
  }
}

File: backend/ZuhooCMS/src/main/java/com/zuhoocms/modules/crm/opportunity/OpportunityService.java
package com.zuhoocms.modules.crm.opportunity;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

public interface OpportunityService {

    OpportunityResponse create(OpportunityRequest request);

    OpportunityResponse createFromLead(Long leadId, OpportunityRequest request);

    Page<OpportunityResponse> listAll(OpportunityStage stage, Long clientId, Long ownerId, Long tagId, String keyword, Pageable pageable);

    OpportunityResponse getById(Long id);

    OpportunityResponse update(Long id, OpportunityRequest request);

    OpportunityResponse changeStage(Long id, ChangeStageRequest request);

    // Preview whether moving this (client-less) Opportunity to WON would auto-link an
    // existing Client. Null means no client-less transition is pending or no match found.
    com.zuhoocms.modules.crm.duplicate.DuplicateMatch previewWonDuplicate(Long id);

    PipelineSummaryResponse getPipelineSummary();

    void delete(Long id);
}
